src/builtin.js

```diff
@@ -929,33 +942,41 @@ Sk.builtin.map = function map (fun, seq) {
         }
         seq = new Sk.builtin.list(combined);
     }
+
     if (!Sk.builtin.checkIterable(seq)) {
         throw new Sk.builtin.TypeError("'" + Sk.abstr.typeName(seq) + "' object is not iterable");
     }
 
-    retval = [];
+    iter = Sk.abstr.iter(seq);
+
+    return (function loopDeLoop(i) {
+        var item = i.tp$iternext();
+
+        if (item === undefined) {
+            return new Sk.builtin.list(retval);
+        }
 
-    for (iter = Sk.abstr.iter(seq), item = iter.tp$iternext();
-         item !== undefined;
-         item = iter.tp$iternext()) {
         if (fun === Sk.builtin.none.none$) {
             if (item instanceof Array) {
                 // With None function and multiple sequences,
                 // map should return a list of tuples
                 item = new Sk.builtin.tuple(item);
             }
             retval.push(item);
-        } else {
-            if (!(item instanceof Array)) {
-                // If there was only one iterable, convert to Javascript
-                // Array for call to apply.
-                item = [item];
-            }
-            retval.push(Sk.misceval.apply(fun, undefined, undefined, undefined, item));
+            return loopDeLoop(iter);
+        }
+
+        if (!(item instanceof Array)) {
+            // If there was only one iterable, convert to Javascript
+            // Array for call to apply.
+            item = [item];
         }
-    }
 
-    return new Sk.builtin.list(retval);
+        return Sk.misceval.chain(Sk.misceval.applyOrSuspend(fun, undefined, undefined, undefined, item), function (result) {
+            retval.push(result);
+            return loopDeLoop(iter);
```

Comment:
This will call into a new stack frame for each element, which will overflow on long sequences.

`Sk.misceval.chain()` isn't like JS promises - if the first `applyOrSuspend()` does *not* suspend, it will call the function straight-out. (And it's not a tail call, so there's no way V8 can save us!) This is a substantial performance boost, which  is why `Sk.misceval.chain()` works like this, but it does mean we need annoying logic to do "loop on this Python iterator but also suspend if it suspends, then resume the loop". Thankfully, this logic is already abstracted into `iterFor()`, so we should use that instead.